### ISSUE TITLE
Testing Improvements Discussion

### ISSUE THREAD

**Issue by armsnyder** [gitlabhq/terraform-provider-gitlab#530]
## Problems

1. I've seen a lot of PRs from first-time contributors get held up on writing new acceptance tests. A common complaint is that they have a difficult time understanding how a new test ought to be written. Most recently this comment: https://github.com/gitlabhq/terraform-provider-gitlab/issues/49#issuecomment-766804318. I'm worried it is impacting our velocity.
2. We have some open issues relating to the acceptance tests, and as these issues aren't being addressed we are copying our bugs every time a new resource is added. (#205 and #434 come to mind.)
3. I've seen bugs slip into test code, likely due to the large amount of boilerplate which contributors don't realize they need to change and reviewers skim through and miss. (e.g. https://github.com/gitlabhq/terraform-provider-gitlab/pull/367#discussion_r484552080 and https://github.com/gitlabhq/terraform-provider-gitlab/pull/367#discussion_r484552569)

Is there something we can do to improve our situation? Let's use this issue to see if we agree/disagree on the problem and maybe brainstorm a bit if it resonates.

## References

* [HashiCorp's testing guide](https://www.terraform.io/docs/extend/testing/index.html)
* [HashiCorp's testing best practices](https://www.terraform.io/docs/extend/best-practices/testing.html)

**Comment by RomanNess** [comment]
I have a side note regarding 1):

In my experience the general quality of tests decreases as the effort to run tests locally increases. Maybe it would help to include a `docker-compose.yml` to run a local dockerized gitlab instance. The GitLab image docs are fairly good, but I've lost around 30min on a recent PR, because I had a port conflict on 8080 within the container. 
One might also think about documenting how to run a single acceptance test from shell. I did it with a GoLand run config since I wasn't sure whether the Makefile supports it.

Obviously both of these measures will not magically increase test quality, but they could help to keep a high quality once it's achieved as new contributors have easier access to running acceptance tests locally.

**Comment by mattkasa** [comment]
I'm not as familiar as I'd like yet with our existing tests, but conceptually I agree with both statements, that reducing boilerplate and effort to add a new test, as well as enabling local test runs would be helpful and decrease the amount of effort to contribute effectively 👍

**Comment by armsnyder** [comment]
Closing in favor of #1008
